protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,405 @@
+                }
+            }
+            p => {
+                if before_circuit {
+                    relayed_multiaddr
+                        .relay_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                } else {
+                    relayed_multiaddr
+                        .dst_addr
+                        .get_or_insert(Multiaddr::empty())
+                        .push(p);
+                }
+            }
+        }
+    }
+
+    Ok(Ok(relayed_multiaddr))
+}
+
+pub struct RelayListener {
+    queued_new_addresses: VecDeque<Multiaddr>,
+    from_behaviour: mpsc::Receiver<ToListenerMsg>,
+    msg_to_behaviour: Option<BoxFuture<'static, Result<(), mpsc::SendError>>>,
+}
+
+impl Unpin for RelayListener {}
+
+impl Stream for RelayListener {
+    type Item =
+        Result<ListenerEvent<Ready<Result<RelayedConnection, RelayError>>, RelayError>, RelayError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(msg) = &mut self.msg_to_behaviour {
+                match Future::poll(msg.as_mut(), cx) {
+                    Poll::Ready(Ok(())) => self.msg_to_behaviour = None,
+                    Poll::Ready(Err(e)) => return Poll::Ready(Some(Err(e.into()))),
+                    Poll::Pending => {}
+                }
+            }
+
+            if let Some(addr) = self.queued_new_addresses.pop_front() {
+                return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr))));
+            }
+
+            match self.from_behaviour.poll_next_unpin(cx) {
+                Poll::Ready(Some(ToListenerMsg::IncomingRelayedConnection {
+                    stream,
+                    src_peer_id,
+                    relay_addr,
+                    relay_peer_id: _,
+                })) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                        upgrade: ready(Ok(stream)),
+                        local_addr: relay_addr.with(Protocol::P2pCircuit),
+                        remote_addr: Protocol::P2p(src_peer_id.into()).into(),
+                    })));
+                }
+                Poll::Ready(Some(ToListenerMsg::Reservation(Ok(Reservation { addrs })))) => {
+                    debug_assert!(
+                        self.queued_new_addresses.is_empty(),
+                        "Assert empty due to previous `pop_front` attempt."
+                    );
+                    // Returned as [`ListenerEvent::NewAddress`] in next iteration of loop.
+                    self.queued_new_addresses = addrs.into();
+                }
+                Poll::Ready(Some(ToListenerMsg::Reservation(Err(())))) => {
+                    return Poll::Ready(Some(Err(RelayError::Reservation)));
+                }
+                Poll::Ready(None) => {
+                    panic!("Expect sender of `from_behaviour` not to be dropped before listener.");
```

Comment:
@mxinden 

Some Getting Code get Panicked. Calling from FFI, happens on WSL as well as, if correctly remember. 
log : 
```
thread '<unnamed>' panicked at 'Expect sender of `from_behaviour` not to be dropped before listener.', C:\Users\PramUkesh\.cargo\git\checkouts\rust-libp2p-cd25f84bb1caf6f6\b122b33\protocols\relay\src\v2\client\transport.rs:325:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

In what case, this might happen ? same code works in times if app is restarted.

Reply:
I'm testing two clients on same pc with different port

Reply:
"Fixed" with https://github.com/libp2p/rust-libp2p/commit/4b90626ab232faacba28139471f7c6fba0b520ab. Note that you will have to handle reconnects manually for now when the connection to the relay closes.